File: tool/gui/src/backend/ai_genre.py
"""Genre tag mapping and template helpers for the AI course generator.

This module keeps the mapping between user-facing genre tags like ``[intro]``
and the internal lesson templates, so both the prompt builder and the UI can
share the same source of truth.
"""
from __future__ import annotations

import re
from typing import Any


GENRE_TEMPLATES: dict[str, dict[str, Any]] = {
    "[intro]": {
        "template": "intro",
        "label": "认识新词",
        "primary_key": "subLessons",
        "description": "通过展示新词、翻译句子和填空引导学生认识生词。",
        "suggested_types": ["showWord", "translateSentence", "fillBlank"],
    },
    "[practice]": {
        "template": "practice",
        "label": "巩固练习",
        "primary_key": "subLessons",
        "description": "通过多轮互动练习巩固词汇和句型。",
        "suggested_types": ["multipleChoice", "translateSentence", "fillBlank", "reorderSentence"],
    },
    "[review]": {
        "template": "review",
        "label": "复习",
        "primary_key": "subLessons",
        "description": "复习已学内容，可混合使用 subLessons 和 stages。",
        "suggested_types": ["multipleChoice", "fillBlank", "translateSentence"],
    },
    "[listening]": {
        "template": "listening",
        "label": "听力训练",
        "primary_key": "listeningPhases",
        "description": "通过听力阶段训练学生的听力理解。",
        "suggested_types": ["listenAndPick", "typeTheWord", "listenOnly"],
    },
    "[reading]": {
        "template": "reading",
        "label": "阅读理解",
        "primary_key": "readingPassage",
        "description": "提供阅读材料并配合阅读理解题目。",
        "suggested_types": ["readingMcq", "readingTrueFalse", "readingShortAnswer"],
    },
    "[mastery]": {
        "template": "mastery",
        "label": "综合测验",
        "primary_key": "stages",
        "description": "单元末综合测验，覆盖多种题型。",
        "suggested_types": ["multipleChoice", "translateSentence", "fillBlank", "multiSelect"],
    },
    "[mixed]": {
        "template": "mixed",
        "label": "混合",
        "primary_key": "",
        "description": "在一个 section 中混合使用多种模板，适合综合课程。",
        "suggested_types": [],
    },
}


TEMPLATE_LABELS: dict[str, str] = {
    "intro": "认识新词",
    "practice": "巩固练习",
    "review": "复习",
    "listening": "听力训练",
    "reading": "阅读理解",
    "mastery": "综合测验",
    "mixed": "混合",
}


TAG_RE = re.compile(r"\[([a-zA-Z]+)\]")


def genre_to_template(genre: str) -> str:
    """Map a genre tag (with or without brackets) to a template name."""
    key = genre.lower().strip()
    if not key.startswith("["):
        key = f"[{key}]"
    meta = GENRE_TEMPLATES.get(key)
    return meta["template"] if meta else "mixed"


def template_label(template: str) -> str:
    """Return the human-readable label for a template."""
    return TEMPLATE_LABELS.get(template, template)


def parse_genre_tag(text: str) -> str | None:
    """Return the first recognized genre tag from ``text``, or None."""
    if not text:
        return None
    for match in TAG_RE.finditer(text):
        tag = f"[{match.group(1).lower()}]"
        if tag in GENRE_TEMPLATES:
            return tag
    return None


def all_genre_tags() -> list[str]:
    """Return all recognized genre tags including brackets."""
    return list(GENRE_TEMPLATES.keys())


def genre_prompt_block() -> str:
    """Return a prompt block that explains available genre tags to the model."""
    lines = ["可用 genre 标签（用户可在主题或额外指令中插入这些标签来指定模板）："]
    for tag, meta in GENRE_TEMPLATES.items():
        lines.append(
            f"{tag} → {meta['label']}（template={meta['template']}，"
            f"主键={meta['primary_key'] or '混合'}）：{meta['description']}"
        )
    lines.append(
        "规则：如果用户在主题或额外指令中插入了 genre 标签，"
        "请按标签为对应单元或课时生成相应模板结构；未标注的部分回退到默认模板。"
    )
    return "\n".join(lines)


def genre_tags_in_text(text: str) -> list[str]:
    """Return all recognized genre tags found in ``text``, in order."""
    found: list[str] = []
    for match in TAG_RE.finditer(text):
        tag = f"[{match.group(1).lower()}]"
        if tag in GENRE_TEMPLATES and tag not in found:
            found.append(tag)
    return found

File: lib/application/ai/ai_genre.dart
// Dart imports:
import 'dart:convert';

/// Genre tag → lesson template metadata. Mirrors
/// `tool/gui/src/backend/ai_genre.py:GENRE_TEMPLATES`.
class GenreMeta {
  const GenreMeta({
    required this.tag,
    required this.template,
    required this.label,
    required this.primaryKey,
    required this.description,
    required this.suggestedTypes,
  });

  final String tag;
  final String template;
  final String label;
  final String primaryKey;
  final String description;
  final List<String> suggestedTypes;
}

const Map<String, GenreMeta> genreTemplates = {
  '[intro]': GenreMeta(
    tag: '[intro]',
    template: 'intro',
    label: 'New Words',
    primaryKey: 'subLessons',
    description:
        'Introduce new words via word display, translation, and fill-in-the-blank.',
    suggestedTypes: ['showWord', 'translateSentence', 'fillBlank'],
  ),
  '[practice]': GenreMeta(
    tag: '[practice]',
    template: 'practice',
    label: 'Practice',
    primaryKey: 'subLessons',
    description:
        'Reinforce vocabulary and sentence patterns through multiple interactions.',
    suggestedTypes: [
      'multipleChoice',
      'translateSentence',
      'fillBlank',
      'reorderSentence',
    ],
  ),
  '[review]': GenreMeta(
    tag: '[review]',
    template: 'review',
    label: 'Review',
    primaryKey: 'subLessons',
    description: 'Review learned content; may mix subLessons and stages.',
    suggestedTypes: ['multipleChoice', 'fillBlank', 'translateSentence'],
  ),
  '[listening]': GenreMeta(
    tag: '[listening]',
    template: 'listening',
    label: 'Listening',
    primaryKey: 'listeningPhases',
    description: 'Train listening comprehension through listening phases.',
    suggestedTypes: ['listenAndPick', 'typeTheWord', 'listenOnly'],
  ),
  '[reading]': GenreMeta(
    tag: '[reading]',
    template: 'reading',
    label: 'Reading',
    primaryKey: 'readingPassage',
    description: 'Provide a reading passage with comprehension questions.',
    suggestedTypes: ['readingMcq', 'readingTrueFalse', 'readingShortAnswer'],
  ),
  '[mastery]': GenreMeta(
    tag: '[mastery]',
    template: 'mastery',
    label: 'Quiz',
    primaryKey: 'stages',
    description: 'End-of-unit quiz covering multiple question types.',
    suggestedTypes: [
      'multipleChoice',
      'translateSentence',
      'fillBlank',
      'multiSelect',
    ],
  ),
  '[mixed]': GenreMeta(
    tag: '[mixed]',
    template: 'mixed',
    label: 'Mixed',
    primaryKey: '',
    description:
        'Combine multiple templates in one section for comprehensive courses.',
    suggestedTypes: [],
  ),
};

const Map<String, String> templateLabels = {
  'intro': 'New Words',
  'practice': 'Practice',
  'review': 'Review',
  'listening': 'Listening',
  'reading': 'Reading',
  'mastery': 'Quiz',
  'mixed': 'Mixed',
};

final RegExp _tagRe = RegExp(r'\[([a-zA-Z]+)\]');

/// Map a genre tag (with or without brackets) to a template name.
String genreToTemplate(String genre) {
  var key = genre.toLowerCase().trim();
  if (!key.startsWith('[')) key = '[$key]';
  final meta = genreTemplates[key];
  return meta?.template ?? 'mixed';
}

/// Return the human-readable label for a template.
String templateLabel(String template) => templateLabels[template] ?? template;

/// Return the first recognized genre tag from [text], or null.
String? parseGenreTag(String? text) {
  if (text == null || text.isEmpty) return null;
  for (final match in _tagRe.allMatches(text)) {
    final tag = '[${match.group(1)!.toLowerCase()}]';
    if (genreTemplates.containsKey(tag)) return tag;
  }
  return null;
}

/// Return all recognized genre tags found in [text], in order.
List<String> genreTagsInText(String? text) {
  final found = <String>[];
  if (text == null || text.isEmpty) return found;
  for (final match in _tagRe.allMatches(text)) {
    final tag = '[${match.group(1)!.toLowerCase()}]';
    if (genreTemplates.containsKey(tag) && !found.contains(tag)) {
      found.add(tag);
    }
  }
  return found;
}

/// All recognized genre tags including brackets.
List<String> allGenreTags() => genreTemplates.keys.toList();

/// Prompt block explaining available genre tags to the model.
String genrePromptBlock() {
  final lines = <String>[
    'Available genre tags (the user may insert these in the topic or extra '
        'instructions to specify a template):',
  ];
  for (final entry in genreTemplates.entries) {
    final meta = entry.value;
    lines.add(
      '${entry.key} → ${meta.label} (template=${meta.template}, '
      'primaryKey=${meta.primaryKey.isEmpty ? "mixed" : meta.primaryKey}): '
      '${meta.description}',
    );
  }
  lines.add(
    'Rule: if the user inserts a genre tag in the topic or extra '
        'instructions, generate the corresponding template structure for the '
        'matching unit or lesson; untagged parts fall back to the default '
        'template.',
  );
  return lines.join('\n');
}

/// Detect first genre tag in spec topic or extra instructions.
String? detectGenreFromSpec({
  required String topic,
  required String extraInstructions,
}) {
  final combined = '$topic $extraInstructions';
  final tags = genreTagsInText(combined);
  return tags.isNotEmpty ? tags.first : null;
}

/// Encode a list of strings as a JSON array string (for DB tags column).
String encodeTags(List<dynamic> tags) => jsonEncode(tags);